tests/test_task_runners.py

```diff
@@ -122,21 +141,85 @@ async def test_wait_captures_exceptions_as_crashed_state(
                 f"{task_runner.concurrency_type} task runners."
             )
 
+        task_run = TaskRun(flow_run_id=uuid4(), task_key="foo", dynamic_key="bar")
+
         async def fake_orchestrate_task_run():
             raise exception
 
-        test_key = uuid4()
-
         async with task_runner.start():
             await task_runner.submit(
+                key=task_run.id,
                 call=partial(fake_orchestrate_task_run),
-                key=test_key,
             )
 
-            state = await (await task_runner.wait(test_key, 5))
+            state = await task_runner.wait(task_run.id, 5)
             assert state is not None, "wait timed out"
             assert isinstance(state, State), "wait should return a state"
-            assert state.name == "Crashed"
+            assert state.type == StateType.CRASHED
+
+    @pytest.mark.parametrize(
+        "exception",
+        [
+            WorkerNodeException(
+                exception_to_raise=KeyboardInterrupt(),
+                expected_exception_type=KilledWorker,
+            ),
+            ValueError("test"),
+        ],
+    )
+    async def test_exception_to_crashed_state_in_flow_run(
+        self, exception, task_runner, monkeypatch
+    ):
+        if task_runner.concurrency_type != TaskConcurrencyType.PARALLEL:
+            pytest.skip(
+                f"This will abort the run for "
+                f"{task_runner.concurrency_type} task runners."
+            )
+
+        # If the type is an unwrapped exception, we can check exception equality
+        # later. If not, we will only check that the exception type raised by the
+        # flow run matches expectations.
+        check_exception_equality = isinstance(exception, BaseException)
+
+        if check_exception_equality:
+            exception_to_raise = exception
+            expected_exception_type = type(exception)
+        else:
+            (exception_to_raise, expected_exception_type) = (
+                exception.exception_to_raise,
+                exception.expected_exception_type,
+            )
```

Comment:
fwiw I think this is pretty complicated and we could just have the parameterization be "raised_exception,state_exception" and avoid all this. You may even be able to use the unittest `ANY` as the exception message to check equality without branching.

Reply:
Agreed. I had something similar originally, and then went back and overthought it 🤔. I added another commit that simplifies the test a bit.